ProjFS.Mac/PrjFSKextTests/KextMockUtilities.hpp

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include <unordered_map>
+#include <unordered_set>
+
+class FunctionTypeCallRecorderBase
+{
+public:
+    virtual void Clear() = 0;
+    virtual ~FunctionTypeCallRecorderBase() {}
+};
+
+template <typename R, typename... ARGS>
+    class FunctionTypeCallRecorder : public FunctionTypeCallRecorderBase
+{
+    typedef R (*FunctionPointerType)(ARGS...);
+    struct FunctionCall
+    {
+        // TODO: add argument values
+        uint64_t callSequenceNumber;
+    };
+    
+    typedef std::unordered_multimap<FunctionPointerType, FunctionCall> RecordedCallMapType;
+
+    RecordedCallMapType recordedCalls;
+    static FunctionTypeCallRecorder functionTypeRegister;
+
+    void RecordFunctionCall(FunctionPointerType function, uint64_t sequenceNumber);
+    bool DidCallFunction(FunctionPointerType function);
+    
+    virtual void Clear() override
+    {
+        this->recordedCalls.clear();
+    }
+    
+    friend class MockCalls;
+};
+
+template <typename R, typename... ARGS>
+    FunctionTypeCallRecorder<R, ARGS...> FunctionTypeCallRecorder<R, ARGS...>::functionTypeRegister;
```

Comment:
I'm a bit confused about how this is working.  

What prevents each each compilation unit that includes KextMockUtilities.hpp from getting its own instance of `functionTypeRegister`? 

(I am assuming that MockVnodeAndMount.cpp and VirtualizationRootTests.mm are each individual units)

Reply:
> What prevents each each compilation unit that includes KextMockUtilities.hpp from getting its own instance of `functionTypeRegister`?

This is as far as I'm aware a guarantee made by the C++ standard - there will only be one instance of a static member for any given template instantiation. I suppose this is comparable to the fact that you don't get duplicate symbol link errors for template function instantiations even if those functions are not inlined. (And also the reason why linkers have to be C++-aware; pure C linkers will not handle C++ object files correctly.)

Reply:
Thanks! I didn't realize that was the case with templates 👍